amplipi/streams.py

```diff
@@ -802,51 +805,77 @@ def connect(self, src):
     self.uuid = uuid_gen()
     portnum = 49494 + int(src)
 
-    # Make the (per-source) config directory
-    src_config_folder = f'{utils.get_folder("config")}/srcs/{src}'
-    os.system(f'mkdir -p {src_config_folder}')
+    # Make the (per-source) config and web directories
+    self.src_config_folder = f'{utils.get_folder("config")}/srcs/{src}'
+    os.system(f'rm -r {self.src_config_folder}')
+    os.system(f'mkdir -p {self.src_config_folder}')
+
+    self.src_web_folder = f'{utils.get_folder("web")}/generated/{src}'
+    os.system(f'rm -r {self.src_web_folder}')
+    os.system(f'mkdir -p {self.src_web_folder}')
+
+    # Make the fifo to be used for commands
+    os.mkfifo(f'{self.src_config_folder}/cmd')
 
-    meta_args = [f'{utils.get_folder("streams")}/dlna_metadata.bash', f'{src_config_folder}']
+    # startup the metadata process and the DLNA process
+    meta_args = [f'{utils.get_folder("streams")}/dlna_meta.py',
+                 f'{self.name}',
+                 f'{self.src_config_folder}/cmd',
+                 f'{self.src_config_folder}/meta.json',
+                 self.src_web_folder, '-d']
     dlna_args = ['gmediarender', '--gstout-audiosink', 'alsasink',
                 '--gstout-audiodevice', utils.real_output_device(src), '--gstout-initial-volume-db',
                 '0.0', '-p', f'{portnum}', '-u', f'{self.uuid}',
-                '-f', f'{self.name}', '--logfile',
-                f'{src_config_folder}/metafifo']
-    self.proc = subprocess.Popen(args=meta_args, preexec_fn=os.setpgrp, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-    self.proc2 = subprocess.Popen(args=dlna_args, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+                '-f', f'{self.name}']
+    self.metadata_proc = subprocess.Popen(args=meta_args)
+    self.dlna_proc = subprocess.Popen(args=dlna_args, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+    self.fifo = os.open(f'{self.src_config_folder}/cmd', os.O_WRONLY, os.O_NONBLOCK) # open the fifo for writing but don't block in case something goes wrong
+
     self._connect(src)
 
   def disconnect(self):
     if self._is_running():
-      os.killpg(os.getpgid(self.proc.pid), signal.SIGKILL)
-      self.proc2.kill()
+      self.metadata_proc.kill()
+      self.dlna_proc.kill()
+      self.fifo.close()
```

Comment:
Is it better to close the fifo first?

Reply:
Or better yet send a close command over the fifo?

Reply:
Recent changes to process handling have added code that looks like this for carefully shutting down processes without leaving any zombies. Here's what the spotify equivalent looks like:

```python3
    if self.proc:
      try:
        self.proc.terminate()
        self.proc.communicate(timeout=3)
      except Exception as e:
        print(f"failed to terminate spotify stream: {e}")
        self.proc.kill()
      self.proc = None
    if self.mpris:
      try:
        del self.mpris
      except Exception:
        pass
      self.mpris = None
``` 

You may want to add some of that idea to this disconnect method.